### PR TITLE
WD-11837 Modify search api to support more filters

### PR DESCRIPTION
**Pull request by immortalcodes**
## Done

- Modified search API to support more filters
- Modifies the database to include a table for authors
- `author_email` acts as a primary key for authors table and foreign key for assets table.
## QA

- Check out this feature branch
- Run the site using the command `dotrun`
- View the site locally in your web browser at: http://0.0.0.0:8017
- Run through the following [QA steps](https://canonical-web-and-design.github.io/practices/workflow/qa-steps.html)
- Use a get request with following URL (add your custom parameters instead of value) `http://0.0.0.0:8017/manager?tag=value&asset_type=value&product_types=value_1&author_email=value&title=value&start_date=value&end_date=value&sf_campg_id=value&language=value&product_types=value_2`

## Issue / Card

Fixes # [WD-11837](https://warthogs.atlassian.net/browse/WD-11837)

## Screenshots

[if relevant, include a screenshot]

[WD-11837]: https://warthogs.atlassian.net/browse/WD-11837?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by petesfrench** [comment]
Need to remove the `query` param from the view https://github.com/canonical/assets.ubuntu.com/blob/WD-11837-modify-search-api-to-support-more-filters/webapp/views.py#L169

File: webapp/dataclass.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+from dataclasses import dataclass
+
+
+@dataclass
+class AssetSearchParams:
+    tag: str = ""
+    asset_type: str = ""
+    product_types: list = None
```
Comment (petesfrench) [inline webapp/dataclass.py:8]:
```suggestion
    tag: str = ""
    asset_type: str = ""
    product_types: list = None
```
Why are we treating the list of regular tags differently to the list of product tags?

Comment (petesfrench) [reply webapp/dataclass.py:8]: Also, I think we should be more consistent with our naming of the tags. From what I understand from the figma, there are two types of tag: 'product' and 'other'. The names should reflect this

File: webapp/models.py [hunk 2]
```diff
@@ -46,18 +46,27 @@ class Token(DateTimeMixin):
...
+    last_name = Column(String, nullable=False, primary_key=True)
+    email = Column(String, nullable=False, unique=True, primary_key=True)
+
+
 class Asset(DateTimeMixin):
     __tablename__ = "asset"
 
     id = Column(Integer, primary_key=True)
     asset_type = Column(String, nullable=True)
     name = Column(String, nullable=True)
```
Comment (petesfrench) [inline webapp/models.py:62]: Has this been replaced by 'title'?

Comment (petesfrench) [reply webapp/models.py:62]: And removed here

**Comment by petesfrench** [comment]
Do we need to test `start_date` and `end_date`?

File: webapp/models.py (hunk 2 above)
Comment (immortalcodes) [reply webapp/models.py:62]: Could you please elaborate this more, I didn't quite get it.
I thought I added the migrations

**Comment by immortalcodes** [comment]
> Do we need to test `start_date` and `end_date`? On the figma it says 'upload date'. I imagine I am lacking some context here

We decided to keep it a range, who could exactly remember the date they uploaded an asset but they could surely know they uploaded it say last month

File: webapp/models.py (hunk 2 above)
Comment (petesfrench) [reply webapp/models.py:62]: You haven't referenced `name` anywhere else in this PR. It seems you have changed it to `title`. We should try to align all our naming.

File: webapp/param_parser.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+def parse_asset_search_params() -> AssetSearchParams:
+    """
+    Parse request arguments and return AssetSearchParams object with defaults.
+    """
+    return AssetSearchParams(
+        tag=request.args.get("tag", "").strip(),
+        asset_type=request.args.get("asset_type", "").strip(),
+        product_types=request.args.getlist("product_types") or [],
+        author_email=request.args.get("author_email", "").strip(),
+        title=request.args.get("title", "").strip(),
```
Comment (petesfrench) [inline webapp/param_parser.py:14]: Here you use `title`

File: webapp/dataclass.py (hunk 1 above)
Comment (petesfrench) [reply webapp/dataclass.py:8]: I guess I am missing some context, but why are we not allowing search for multiple tags?

**Comment by petesfrench** [comment]
I am getting some strange behavior when searching by dates.
Here is my current DB content:
```
 id |          created           |                data                 |                      file_path                      | deprecated |   asset_type    |      name       |  google_drive_link   | salesforce_campaign_id | language |          updated           |      author_email      
----+----------------------------+-------------------------------------+-----------------------------------------------------+------------+-----------------+-----------------+----------------------+------------------------+----------+----------------------------+------------------------
  6 | 2024-10-11 17:04:13.027803 | {"image": true, "optimized": false} | 62a77319-@pleia2.png                                | f          | case-study      | fgdsfgdgsdgdfs  |                      |                        |          | 2024-10-11 17:04:13.031327 | john.doe@example.com
  7 | 2024-10-11 17:17:58.361226 | {"optimized": false}                | da39a3ee                                            | f          | datasheets      |                 |                      |                        |          | 2024-10-11 17:17:58.364714 | jane.smith@example.com
  8 | 2024-10-13 10:37:13.213971 | {"image": true, "optimized": false} | 2a9c0c18-sofa.png                                   | f          | icon            | blabla          |                      |                        |          | 2024-10-13 10:37:13.216478 | jane.smith@example.com
  9 | 2024-10-13 11:08:10.94565  | {"image": true, "optimized": false} | 426fa7cc-pexels-bella-white-201200-635279.jpg       | f          | icon            | nameymcnameface |                      |                        |          | 2024-10-13 11:08:10.948364 | jane.smith@example.com
 11 | 2024-10-13 11:36:02.432274 | {"optimized": false, "image": true} | 33b80177-enhanced_image.png                         | f          | ebook           | pop pop         | www.linkylink        |                        | chinese  | 2024-10-13 11:36:02.435712 | jane.smith@example.com
 12 | 2024-10-13 12:20:04.278824 | {"optimized": false, "image": true} | 9a62ef78-@shaakunthala.png                          | f          | guides          | ooooooook       | www.linkylink.coooom |                        | Bengali  | 2024-10-13 12:20:04.279746 | jane.smith@example.com
  1 | 2024-10-11 07:52:11.161793 | {"image": true, "optimized": false} | ff2dfa4f-ubuntu-summit-2023-background-logo (1).svg | f          | image           | test            | https://linky.link   | 12342                  | en2      | 2024-10-11 07:52:11.165768 | john.doe@example.com
  2 | 2024-10-11 08:04:08.792757 | {"image": true, "optimized": false} | 4d41f82b-circle-of-friends.png                      | f          | image           | test            | https://linky.link   | 12342                  | en2      | 2024-10-11 08:04:08.794934 | john.doe@example.com
  3 | 2024-10-11 10:14:47.603193 | {"image": true, "optimized": false} | 685bc1a3-flags.png                                  | f          | UType           | test            | ULink                | USFCI                  | ULang    | 2024-10-11 11:34:58.203218 | john.doe@example.com
  4 | 2024-10-11 15:22:54.377545 | {"image": true, "optimized": false} | 2133f0e1-Ubuntu-Logo-Dark.svg                       | f          | image           | test            | https://linky.link   | 12342                  | en2      | 2024-10-11 15:22:54.381708 | john.doe@example.com
  5 | 2024-10-11 17:00:52.153843 | {"image": true, "optimized": false} | 714c5293-ubuntu-mate-circle-icon.png                | f          | explainer-video | test            |                      |                        |          | 2024-10-11 17:00:52.155347 | john.doe@example.com
 10 | 2024-10-13 11:33:56.650573 | {"image": true, "optimized": false} | 0ca5fef2-pexels-veeterzy-38136.jpg                  | f          | image           | hipposarecool   | www.totallyreal.com  |                        | en2      | 2024-10-13 11:33:56.652037 | jane.smith@example.com
(12 rows)
```
Here is my query in postman: `http://0.0.0.0:8017/manager?tag=test&asset_type=image&author_email=john.doe@example.com&product_types=product2&product_types=support&title=test&language=en2&start_data=2024-10-12&end_data=2024-10-15`
It returns `2133f0e1-Ubuntu-Logo-Dark.svg` and `4d41f82b-circle-of-friends.png`.
Here is the same query but in SQL:
```
SELECT a.*
FROM asset a
JOIN asset_tag_association ata ON a.id = ata.asset_id
JOIN asset_product_association apa ON a.id = apa.asset_id
WHERE a.author_email = 'john.doe@example.com'
  AND a.asset_type = 'image'
  AND a.name = 'test'
  AND a.language = 'en2'
  AND ata.tag_name = 'test'
  AND apa.product_name IN ('product2', 'support')
  AND a.created BETWEEN '2024-10-12' AND '2024-10-15';
  ```
  And the result:
  ```
   id | created | data | file_path | deprecated | asset_type | name | google_drive_link | salesforce_campaign_id | language | updated | author_email 
----+---------+------+-----------+------------+------------+------+-------------------+------------------------+----------+---------+--------------
(0 rows)
```

**Comment by immortalcodes** [comment]
> 4 | 2024-10-11 15:22:54.377545 | {"image": true, "optimized": false} | 2133f0e1-Ubuntu-Logo-Dark.svg                       | f          | image           | test            | https://linky.link   | 12342                  | en2      | 2024-10-11 15:22:54.381708 | john.doe@example.com

I am checking for `update_date` and the SQL query is for created date

File: webapp/param_parser.py (hunk 3 above)
Comment (immortalcodes) [reply webapp/param_parser.py:14]: Oh, alright so the thing is that is the Figma has disparity in it, it says it `asset name` during creation and `Title` during search.

I kept like that for sake of UI and frontend but we can definitely make it consistent.
So what shall we keep?

Comment (immortalcodes) [reply webapp/param_parser.py:14]: image

Comment (petesfrench) [reply webapp/param_parser.py:14]: I would make it 'title' everywhere

**Review by petesfrench — COMMENTED** [review]
LGTM, just align the use of `name` and `title` and you can merge

**Pull request merged**